velox/substrait/SubstraitParser.cpp

```diff
@@ -60,6 +60,11 @@ std::shared_ptr<SubstraitParser::SubstraitType> SubstraitParser::parseType(
       nullability = substraitType.string().nullability();
       break;
     }
+    case ::substrait::Type::KindCase::kUserDefined: {
```

Comment:
Velox doesn't support user-defined types. Mapping these to UNKNOWN type is incorrect. Let's throw instead.

```
VELOX_UNSUPPORTED("Velox doesn't support user-defined types.")
```

Reply:
Hi @mbasmanova , I found that velox treat NULL as UNKNOWN type,  is this right?

Reply:
NULL literal whose type is not known is represented using UNKNOWN type.

Reply:
so, for the query `SELECT NULL` , I would like to convert it into `kUserDefined` in substrait, since we don't know the type. Or could you please help give an example query that could use the null literal in expr that has the type? Any suggestions about this?

Reply:
The problem with using kUserDefined is that it can represent something other than UNKNOWN type. We would need to add some extra checks to ensure that kUserDefined is indeed represents the UNKNOWN type.

What's the motivation for this PR in the first place? What didn't work before and is working now?

Reply:
My reading of the Substrait spec suggests that to introduce a user-defined type, one needs to create an extension: https://substrait.io/types/type_classes/#user-defined-types

Reply:
We just want to add the property nullability in the type. And so that can check it in our internal code and do some other work.

Reply:
> We just want to add the property nullability in the type. And so that can check it in our internal code and do some other work.

Can you explain a bit more? Since types are always non-null, what kind of functionality are you looking to build based on types nullability?

Reply:
Yes, after talk with you, we just need set it to `::substrait::Type_Nullability_NULLABILITY_NULLABLE`.

Reply:
> My reading of the Substrait spec suggests that to introduce a user-defined type, one needs to create an extension: https://substrait.io/types/type_classes/#user-defined-types

I will try to add an extension and limit it to unknown type.

Reply:
> > We just want to add the property nullability in the type. And so that can check it in our internal code and do some other work.
> 
> Can you explain a bit more? Since types are always non-null, what kind of functionality are you looking to build based on types nullability?

This is mainly because our codegen will call different functions according to whether the data has null or not

Reply:
> This is mainly because our codegen will call different functions according to whether the data has null or not

That would make sense, but I wonder how would the information in the plan help with that. What's the end-to-end flow? Do you execute the query plan and then look at the data to see if it has nulls and decide which codegen path to take?

Reply:
Yes,
The flow is : velox plan ->substrait plan->modularSQL IR->codegen and the IR needs the null information of data to call codegen to generate different code paths.

Reply:
@ZJie1 Thank you for clarifying. A followup question: 

> IR needs the null information of data

Can you be a bit more specific about what information and what data do you refer to? The data itself is not available in the plan as it becomes available only during execution. The plan does contain the data types, but all these are non-null . What other information does the codegen use from the plan?

Reply:
Actually I think we need the table schema information, while the compile and runtime part are responsible by my colleagues, so I'm not particularly clear, sorry about this.

Reply:
@ZJie1 

> while the compile and runtime part are responsible by my colleagues, so I'm not particularly clear, sorry about this.

Let's pull in folks who have the necessary context so we can understand these changes fully.

Reply:
@mbasmanova The null information is more like a hint to ModularSQL IR. If the IR is able to know whether the data has nulls in advance it can generate more efficient code. This is more suitable for test plan with ValuesNode, if we know the info for data range and nullable it can chose the best hash join type for codegen.

Reply:
@xieqi Thank you for clarifying. This makes sense. Velox plan doesn't carry this kind of information though. Also, it is not clear to me if Substrait allows for expressing these statistics about the data. I would imagine these do not belong to data type, but rather need to be specified separately.